File: leetcode/Python/longest-consecutive-sequence.py
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        us, sol, n , s, t, i = set(nums), 0, 0, 0, 0, 0
        if not nums:
            return 0
        while i<len(nums):
            if s==0:
                t = nums[i]
            if t in us:
                us.remove(t)
                if s==0:
                    n, s, t = 1, 1, t-1
                elif s==1:
                    n,t = n+1,t-1
                elif s==2:
                    n,t = n+1,t+1
                sol=max(sol,n)
            else:
                if s>0:
                    s=(s+1)%3
                    if s==2:
                        t=nums[i]+1
                if s==0:
                    i+=1
        return sol
                    
                

File: leetcode/C++/longest-consecutive-sequence.cpp
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        int sol=0,n=0;
        unordered_set<int> us(nums.begin(),nums.end());
        if(nums.empty()) return 0;
        for(int i=0,t=0,s=0; i<nums.size();){
            if(s==0){
                if(us.find(nums[i])==us.end()) ++i;
                else {n=s=1; t=nums[i]; us.erase(t);}
            }
            else if(s==1){
                if(us.find(--t)==us.end()) {t=nums[i]; s=2;}
                else {++n; us.erase(t);}
            }
            else {
                if(us.find(++t)==us.end()) {s=0; ++i;}
                else {++n; us.erase(t);} 
            }
            sol=max(sol,n);
        }
        return sol;
    }
};